source/slang/slang-ir-explicit-global-context.cpp

```diff
@@ -0,0 +1,523 @@
+        SLANG_UNEXPECTED("no outer func at use site for global");
+        UNREACHABLE_RETURN(nullptr);
+    }
+
+    IRInst* findOrCreateContextPtrForFunc(IRFunc* func)
+    {
+        // At this point we are being asked to either find or
+        // produce a context pointer for use inside `func`.
+        //
+        // If we already created such a pointer (perhaps because
+        // `func` is an entry point), then we are home free.
+        //
+        if( auto found = m_mapFuncToContextPtr.TryGetValue(func) )
+        {
+            return *found;
+        }
+
+        // Otherwise, we are going to need to introduce an
+        // explicit parameter to `func` to represent the
+        // context.
+        //
+        IRBuilder builder(m_sharedBuilder);
+
+        // We can safely assume that `func` has a body, because
+        // otherwise we wouldn't be getting a request for the
+        // context pointer value to use in its body.
+        //
+        auto firstBlock = func->getFirstBlock();
+        SLANG_ASSERT(firstBlock);
+
+        // We create a new parameter at the end of the parameter
+        // list for `func`, with a type of `KernelContext*`.
+        //
+        IRParam* contextParam = builder.createParam(m_contextStructPtrType);
+        addKernelContextNameHint(contextParam);
+        contextParam->insertBefore(firstBlock->getFirstOrdinaryInst());
+
+        // The new parameter can be registerd as the context value
+        // to be used for `func` right away.
+        //
+        // Note: we register the value *before* modifying locations
+        // that call `func` to protect against a possible infinite-recursion
+        // situation if `func` is recursive along some path.
+        //
+        m_mapFuncToContextPtr.Add(func, contextParam);
+
+        // Any code that calls `func` now needs to be updated to pass
+        // the context parameter.
+        //
+        // TODO: There is an issue here if `func` might be called
+        // dynamically, through something like a witness table.
+        //
+        List<IRUse*> uses;
+        for( auto use = func->firstUse; use; use = use->nextUse )
+        {
+            // We will only fix up calls to `func`, and ignore
+            // other operations that might refer to it.
+            //
+            // TODO: We need to allow things like decorations that might
+            // refer to `func`, but this logic is also going to
+            // ignore things like witness tables that refer to `func`,
+            // or operations that pass `func` as a function pointer
+            // to a higher-order function.
+            //
+            auto call = as<IRCall>(use->getUser());
+            if(!call)
+                continue;
+
+            // We are going to construct a new call to `func`
+            // that has all of the arguments of the original call...
+            //
+            UInt originalArgCount = call->getArgCount();
+            List<IRInst*> args;
```

Comment:
At some point we should add an overload of `emitCallInst` so we can start to use `ShortList`.